garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,282 @@
+
+                td_error = tf.stop_gradient(target_q_values) - q_selected
+                loss = tf.square(td_error)
+                self._loss = tf.reduce_mean(loss)
+
+            with tf.name_scope("optimize_ops"):
+                optimizer = self.qf_optimizer(self.qf_lr)
+                if self.grad_norm_clipping is not None:
+                    gradients = optimizer.compute_gradients(
+                        self._loss, var_list=self.qf.get_trainable_vars())
+                    for i, (grad, var) in enumerate(gradients):
+                        if grad is not None:
+                            gradients[i] = (tf.clip_by_norm(
+                                grad, self.grad_norm_clipping), var)
+                    self._optimize_loss = optimizer.apply_gradients(gradients)
+                else:
+                    self._optimize_loss = optimizer.minimize(
+                        self._loss, var_list=self.qf.get_trainable_vars())
+
+    @overrides
+    def train(self, sess=None):
+        """
+        Train the network.
+
+        A tf.Session can be provided, or will be created otherwise.
+        """
+        created_session = True if sess is None else False
+        if sess is None:
+            self.sess = tf.Session()
+            self.sess.__enter__()
+        else:
+            self.sess = sess
+
+        self.sess.run(tf.global_variables_initializer())
+
+        episode_rewards = []
+        episode_qf_losses = []
+        episode_length = []
+        ts = 0
+
+        self.sess.run(self._qf_update_ops, feed_dict=dict())
+
+        obs = self.env.reset()
+
+        # Baselines atari wrappers use a LazyFrame object to store the
+        # observation. The way to unpack it is to unpack obs as numpy array.
+        # For reference, see
+        # https://github.com/openai/baselines/blob/1259f6ab25c6f7261e33c
+        # 4c3b92df869188f9260/baselines/common/atari_wrappers.py#L202
+        if self.use_atari_wrappers:
+            obs = np.asarray(obs)
+        episode_rewards.append(0.)
+
+        for itr in range(self.num_timesteps):
+            with logger.prefix('Iteration #%d | ' % itr):
+                # normalize observation for pixel environment
+                if len(self.env.spec.observation_space.shape) == 3:
+                    obs_scaled = obs.astype(np.float32) / 255.0
+                else:
+                    obs_scaled = obs
+
+                if self.es:
+                    action, _ = self.es.get_action(itr, obs_scaled,
+                                                   self.policy)
+                else:
+                    action, _ = self.policy.get_action(obs_scaled)
+
+                next_obs, reward, done, env_info = self.env.step(action)
+
+                if self.use_atari_wrappers:
+                    next_obs = np.asarray(next_obs)
+                self.replay_buffer.add_transitions(
+                    observation=[obs],
```

Comment:
Could we add normalized obs and next_obs to replay buffer?

Reply:
Normalized obs and next_obs will be float. To save memory, we better store uint8 and only normalize them when needed, so I think we could, but we shouldn't.

Reply:
i c. thanks for explaining it.